README.md
# Simple PHP Redirect

This is a _really_ simple project. And as it's only my second
foray into PHP, I doubt it's very well coded.

The application is just a URI redirect, so for example if I
wanted to link to the Cyotek Blog (which is currently at
<https://www.cyotek.com/blog>) from an application I could
create an entry named "blog" pointing to that URL and then my
application would call `https://go.cyotek.com?id=blog`. If this
was later changed to <https://blog.cyotek.com/> (which it was!),
I can update the entry in the database and existing code will
find the new location.

Of course, if I _did_ do that, I'd still have put some sort of
redirect in place which makes this project a bit moot for URL's
that you have full control over. However if you make use of
external services such as PayPal then this project could be more
useful. Another slight benefit is that URL's should be slightly
shorter, although this isn't really the goal of this project.

## Requirements

* PHP with PDO and SQLite support
* Raw access to the database as there's no management interface
  to change settings or add new URI's.
* [Composer][composer] is required for
  downloading vendor packages

## Setup

Create a SQLite database named `redirects.db` and apply
`redirects.db.sql` to this database.

Populate the `Url` table with URL's.

There are two configuration settings in the `Config` table,
**Title** and **HomePage**. Use these to configure the
information displayed when viewing index.php without providing a
redirect ID, which is slightly more friendly than display an
"not set" error.

## Calling the redirect

Call `index.php` with a `id` query string parameter that
contains either the numeric ID or slug of the URL to redirect.

### Parameter injection

If your URL entry includes `{name}` sequences, then the redirect
will try and replace these with query string parameters.

For example, if you had a URL entry named `docs` set with the
URL pattern `https://docs.cyotek.com/cyowcopy/{version}/`, you
could call it as `?id=docs&version=1.2`, resulting in a final
URL of `https://docs.cyotek.com/cyowcopy/1.2/`. Missing
parameters will be substituted with blank values.

## GDPR and Privacy

If the `AnonymizeAddresses` configuration property is set to
`true`, IP addresses will be automatically anonymised.

## Using the management pages

There are several basic overview pages, `activity.php`,
`error.php` and `list.php`. These are "protected" via the use of
an API key, which can be supplied by either the `ApiKey` header
or `apikey` query string parameter.

To configure the API key, change the value of the **ApiKey**
entry in the `Config` table. Note that if the key is left blank,
the management pages will not load.

> There are no configuration pages for adding, removing or
> updating URL entries - you'd need to dig into the database to
> do that. I do plan on adding them at some point, but at the
> same time I don't want to worry about injection attacks or
> weak security given my beginners knowledge of PHP.

Two additional pages exist, also protected via the API key. The
first, `config.php` simply lists the contents of the `Config`
table. The second, `migrate.php` does a basic check of all IP
addresses in the database, and anonymises them (if the
`AnonymizeAddresses` parameter is set)

[composer]: https://getcomposer.org/


CHANGELOG.md
# Change Log

## 1.5.0

### Added

* Added navigation link for URLs
* Added ability to edit each link

### Fixed

* URL list sorting was case sensitive

## 1.4.0

### Added

* Added rudimentary Add URL form to the redirects summary page

### Fixed

* Migration didn't anonymise addresses in the `Error` table
* Added `die` to error routines to avoid dropping into main
  script if the caller wasn't using conditions

## 1.3.0

### Added

* Added new migration page to anonymise existing data
* Added new configuration viewer page

### Fixed

* Fixed syntax error in `UPGRADE_1_1_0_TO_1_2_0.sql`

## 1.2.0

### Added

* Added new `AnonymizeAddresses` setting. When set to `true`
  (default), IP addresses will be anonymised to a /24 subnet
  (IPv4) or a /64 subnet (IPv6)

## 1.1.0

### Added

* Added ability to record 400 and 404 errors to the database
* Added new `errors.php` management page for viewing the error
  log
* Added version information to page footers

### Fixed

* Calling `index.php` without providing an redirect ID now
  returns `200` and displays a slightly more friendly/usable
  page defined using the **Title** and **HomePage**
  configuration values.

## 1.0.0

* Initial release


src/list.php
<?php require_once __DIR__ . '/includes/functions.php'; ?>
<?php
  if(!is_valid_apikey())
  {
    write_forbidden('Redirects Summary', 'Invalid API key.');
  }
  else
  {
    $title='';
    $slug = '';
    $url = '';

    if(isset($_POST['submit']))
    {
      $title = $_POST['title'];
      $slug = $_POST['slug'];
      $url = $_POST['url'];

      if(!is_null_or_empty($title) && !is_null_or_empty($slug) && !is_null_or_empty($url))
      {
        $id = $_POST['id'];

        if(is_numeric($id))
        {
          update_redirect($id, $_POST['enabled'] === 'on' ? 1 : 0, $title, $slug, $url);

          $title='';
          $slug = '';
          $url = '';
        }
        else
        {
          add_redirect($title, $slug, $url);
        }
      }
    }

    $db = get_database_connection();
    $result = $db->query('SELECT [Id], [Enabled], [Title], [Slug], [Url], [LastHit], [HitCount] FROM [UrlSummary] ORDER BY [Title] COLLATE NOCASE ASC');
?>

<?php write_header_html('Redirects Summary') ?>
<div class="wrapper wrapper-90">
  <h1>Redirects Summary</h1>

  <div class="table">
    <div>
      <div>ID</div>
      <div>Enabled?</div>
      <div>Title</div>
      <div>Slug</div>
      <div>Url</div>
      <div>Hits</div>
      <div>Last Hit</div>
      <div>&nbsp;</div>
    </div>
    <form action="list.php?apikey=<?php echo get_api_key() ?>"
      method="post">
      <div>&nbsp;</div>
      <div>&nbsp;</div>
      <div>
        <input id="title" name="title" type="text" maxlength="255"
          value="<?php echo $title ?>" class="title-editor" />
      </div>
      <div>
        <input id="slug" name="slug" type="text" maxlength="255"
          value="<?php echo $slug ?>" class="slug-editor" />
      </div>
      <div>
        <input id="url" name="url" type="text" maxlength="255"
          value="<?php echo $url ?>" class="url-editor" />
      </div>
      <div>&nbsp;</div>
      <div>&nbsp;</div>
      <div>
        <button name="submit" type="submit">Add</button>
      </div>
    </form>
    <?php foreach($result as $row) { ?>
    <form action="list.php?apikey=<?php echo get_api_key() ?>"
      method="post">
      <div class="right"><input id="id" name="id" type="hidden"
          value="<?php echo($row['Id']) ?>" /><?php echo($row['Id']) ?>
      </div>
      <div><input id="enabled" name="enabled" type="checkbox"
          <?php echo($row['Enabled'] ? 'checked' : '') ?> /></div>
      <div>
        <input id="title" name="title" type="text" maxlength="255"
          value="<?php echo($row['Title']) ?>" class="title-editor" />
      </div>
      <div>
        <input id="slug" name="slug" type="text" maxlength="255"
          value="<?php echo($row['Slug']) ?>" class="slug-editor" />
      </div>
      <div>
        <input id="url" name="url" type="text" maxlength="255"
          value="<?php echo $row['Url'] ?>" class="url-editor" />
        <a href="<?php echo $row['Url'] ?>" title="Visit <?php echo $row['Url'] ?>" target="_blank">&#11179;</a>
      </div>
      <div class="right"><?php echo($row['HitCount']) ?></div>
      <div>
        <?php echo($row['LastHit'] != null ? date('Y-m-d H:i', $row['LastHit']) : 'Never') ?>
      </div>
      <div><button name="submit" type="submit">Update</button></div>
    </form>
    <?php } ?>
  </div>
</div>
<?php write_footer_html() ?>

<?php
   $db = null;
    }
?>


src/error.php
<?php require_once __DIR__ . '/includes/functions.php'; ?>
<?php
  if(!is_valid_apikey())
  {
    write_forbidden('Errors', 'Invalid API key.');
  }
  else
  {
    $db = get_database_connection();
    $result = $db->query('SELECT [Url], [Timestamp], [IpAddress], [StatusCode] FROM [Error] ORDER BY [Timestamp] DESC LIMIT 1000');
?>

<?php write_header_html('Errors') ?>
<div class="wrapper wrapper-90">
  <h1>Errors</h1>
  <table>
    <thead>
      <tr>
        <th>Status Code</th>
        <th>Url</th>
        <th>Timestamp</th>
        <th>Client</th>
      </tr>
    </thead>
    <tbody>
    <?php foreach($result as $row) { ?>
      <tr>
        <td><?php echo($row['StatusCode']) ?></td>
        <td><?php echo($row['Url']) ?></td>
        <td><?php echo(date('Y-m-d H:i:s', $row['Timestamp'])) ?></td>
        <td><?php echo($row['IpAddress']) ?></td>
      </tr>
      <?php } ?>
    </tbody>
  </table>
</div>
<?php write_footer_html() ?>

<?php
   $db = null;
    }
?>


src/migrate.php
<?php require_once __DIR__ . '/includes/functions.php'; ?>
<?php
use geertw\IpAnonymizer\IpAnonymizer;

function anonymise_table($tableName)
{
  $db = get_database_connection();
  $result = $db->query('SELECT [Id], [IpAddress] FROM [' . $tableName . '];');
  $total = 0;

  $sql = 'UPDATE [' . $tableName . '] '
                . 'SET [IpAddress] = :address '
                . 'WHERE [Id] = :id';

  $stmt = $db->prepare($sql);

  foreach($result as $row)
  {
    $id = $row['Id'];
    $address =  $row['IpAddress'];
    $newAddress =  IpAnonymizer::anonymizeIp($address);
    
    if($address != $newAddress)
    {
      $stmt->bindValue(':id', $id);
      $stmt->bindValue(':address', $newAddress);
      $stmt->execute();
      $total += $stmt->rowCount();
    }    
  }

  $db = null; 

  return $total;
}

function update_system_version()
{
  update_configuration_field('SystemVersion', '1.5.0');
}

function update_configuration_field($name, $value)
{
  $db = get_database_connection();
  
  $sql = 'UPDATE [Config] '
                . 'SET [Value] = :value '
                . 'WHERE [Name] = :name';

  $stmt = $db->prepare($sql);
  $stmt->bindValue(':name', $name);
  $stmt->bindValue(':value', $value);
  $stmt->execute();
}

if(!is_valid_apikey())
  {
    write_forbidden('Migrate Data', 'Invalid API key.');
  }
  else
  {
?>

<?php write_header_html('Migrate Data') ?>
<div class='wrapper wrapper-90'>
  <h1>Data Migration</h1>

  <?php 

if(should_anonymise_ip_address())
{
  $total = anonymise_table('Activity');
  $total += anonymise_table('Error');

  echo '<p>' . $total . ' records updated.</p>';
}
else
{
  echo '<p>No migration performed.</p>';
}
 ?>

 <?php update_system_version(); ?>

</div>
<?php write_footer_html() ?>

<?php
    }
?>


src/activity.php
<?php require_once __DIR__ . '/includes/functions.php'; ?>
<?php
  if(!is_valid_apikey())
  {
    write_forbidden('Activity', 'Invalid API key.');
  }
  else
  {
    $db = get_database_connection();
    $result = $db->query('SELECT [Title], [Url], [Timestamp], [IpAddress] FROM [ActivitySummary] ORDER BY [Timestamp] DESC LIMIT 1000');
?>

<?php write_header_html('Activity') ?>
<div class="wrapper wrapper-90">
  <h1>Activity</h1>
  <table>
    <thead>
      <tr>
        <th>Title</th>
        <th>Url</th>
        <th>Timestamp</th>
        <th>Client</th>
      </tr>
    </thead>
    <tbody>
    <?php foreach($result as $row) { ?>
      <tr>
        <td><?php echo($row['Title']) ?></td>
        <td><?php echo($row['Url']) ?></td>
        <td><?php echo(date('Y-m-d H:i:s', $row['Timestamp'])) ?></td>
        <td><?php echo($row['IpAddress']) ?></td>
      </tr>
      <?php } ?>
    </tbody>
  </table>
</div>
<?php write_footer_html() ?>

<?php
   $db = null;
    }
?>


src/includes/functions.php
<?php
use geertw\IpAnonymizer\IpAnonymizer;
require 'vendor/autoload.php';

function should_anonymise_ip_address()
{
  return filter_var(get_config_value('AnonymizeAddresses'), FILTER_VALIDATE_BOOLEAN);
}

function get_redirect_id()
{
  if(isset($_GET['id']))
  {
    $id = $_GET['id'];
  }
  //else if(isset($_SERVER['QUERY_STRING']))
  //{
  //  $id = $_SERVER['QUERY_STRING'];
  //}
  else
  {
    $id = -1;
  }

  if(!is_numeric($id))
  {
    $id = get_id_from_slug($id);
  }

  return $id;
}

function is_test_mode()
{
  if(isset($_GET['test']))
  {
    $value = (int)$_GET['test'];
    
    $result = $value != 0;
  }
  else
  {
    $result = false;
  }

  return $result;
}

function write_forbidden($title, $reason)
{
  $html = get_html_document($title, $reason);

  write_error(403);

  header('HTTP/1.1 403 Forbidden');
  header('Content-Type: text/html');
  header('Content-Length: ' . strlen($html));

  echo $html;

  die();
}

function write_bad_request($title, $reason)
{
  $html = get_html_document($title, $reason);

  write_error(400);

  header('HTTP/1.1 400 Bad Request');
  header('Content-Type: text/html');
  header('Content-Length: ' . strlen($html));

  echo $html;

  die();
}

function write_not_found()
{
  $html = get_html_document('Not found', 'The requested resource cannot be found.');

  write_error(404);

  header('HTTP/1.1 404 Not Found');
  header('Content-Type: text/html');
  header('Content-Length: ' . strlen($html));

  echo $html;

  die();
}

function write_redirect($url, $includeLocationHeader)
{
  $html = get_html_document('Found', '<p>The requested resource can be found <a href="' . $url . '">here</a>.</p>');

  header('HTTP/1.1 302 Found');
  if($includeLocationHeader)
  {
    header('Location: ' . $url);
  }
  header('Content-Type: text/html');
  header('Content-Length: ' . strlen($html));

  echo $html;

  die();
}

function get_id_from_slug($slug)
{
  try
  {
    $db = get_database_connection();
    $id = -1;

    $sql = $db->prepare('SELECT [Id] FROM [Url] WHERE [Slug] = ?');

    if($sql->execute(array($slug)))
    {
      $row = $sql->fetch();
      
      if($row)
      {
        $id = $row['Id'];
      }
    }
    
    $db = null;
  }
  catch(Exception $e)
  {
    $id = -1;
  }

  return $id;
}

function get_redirect_uri($id)
{
  try
  {
    $db = get_database_connection();
    $uri = null;

    $sql = $db->prepare('SELECT [Enabled], [Url] FROM [Url] WHERE [Id] = ?');

    if($sql->execute(array($id)))
    {
      $row = $sql->fetch();
      
      if($row && $row['Enabled'] != 0)
      {
        $uri = $row['Url'];

        if(strpos($uri, '{') >= 0)
        {
          $uri = replace_parameters($uri);
        }
      }
    }
    
    $db = null;
  }
  catch(Exception $e)
  {
    $uri = null;
  }

  return $uri;
}

function replace_parameters($value)
{
  return preg_replace_callback('|\{(.*?)\}|',
    function ($matches) 
    {
      $param = $matches[1];

      if(isset($_GET[$param]))
      {
        $result = $_GET[$param];
      }
      else
      {
        $result = '';
      }
    
      return $result;
    },
    $value);
}

function increment_hit_count($id, $url)
{
  try
  {
    $db = get_database_connection();

    $timestamp = time();
    $address = get_client_ip_address();

    $sql = $db->prepare('INSERT INTO [Activity] ([UrlId], [Timestamp], [IpAddress], [Url]) VALUES (?, ?, ?, ?)');
    $sql->execute(array($id, $timestamp, $address, $url));

    $db = null;
  }
  catch(Exception $e)
  {
    // TODO: Do something
  }
}

function get_client_ip_address()
{
  if(isset($_SERVER['HTTP_X_FORWARDED_FOR']))
  {
    $address = $_SERVER['HTTP_X_FORWARDED_FOR'];
  }
  else 
  {
    $address = $_SERVER['REMOTE_ADDR'];
  }

  if(should_anonymise_ip_address())
  {
    $address = IpAnonymizer::anonymizeIp($address);
  }

  return $address;
}

function get_api_key()
{
  if(isset($_SERVER['HTTP_APIKEY']))
  {
    $key = $_SERVER['HTTP_APIKEY'];
  }
  elseif(isset($_GET['apikey']))
  {
    $key = $_GET['apikey'];
  }
  else 
  {
    $key = null;
  }

  return $key;
}

function get_config_value($name)
{
  try
  {
    $db = get_database_connection();

    $result = null;

    $sql = $db->prepare('SELECT [Value] FROM [Config] WHERE [Name] = ?');

    if($sql->execute(array($name)))
    {
      $row = $sql->fetch();
      
      if($row)
      {
        $result = $row['Value'];
      }
    }
    
    $db = null;
  }
  catch(Exception $e)
  {
    $result = null;
  }

  return $result;
}

function is_valid_apikey()
{
  $key = get_api_key();

  if(isset($key) && $key !== '')
  {
    $result = get_config_value('ApiKey') === $key;
  }
  else
  {
    $result = false;
  }

  return $result;
}

function get_title()
{
  return get_config_value('Title');
}

function get_home_page_url()
{
  return get_config_value('HomePage');
}

function get_system_version()
{
  return get_config_value('SystemVersion');
}

function get_schema_version()
{
  try
  {
    $db = get_database_connection();
    
    $version = null;
    $result = $db->query('SELECT [SchemaVersion] FROM [SystemVersion];');

    if($result)
    {
      $row = $result->fetch();

      if($row)
      {
        $version = $row['SchemaVersion'];
      }
    }

    $db = null;
  }
  catch(Exception $e)
  {
    $version = null;
  }

  return $version;}

function get_database_connection()
{
  $connection = new PDO('sqlite:data/redirects.db');

  $connection->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
  $connection->setAttribute(PDO::ATTR_EMULATE_PREPARES, false);

  return $connection;
}

function get_html_document($title, $html)
{
  return get_header_html($title) . '<div class="wrapper">' . $html . '</div>' . get_footer_html();
}

function write_error($code)
{
  try
  {
    $db = get_database_connection();

    $url = $_SERVER['REQUEST_URI'];
    $timestamp = time();
    $address = get_client_ip_address();

    $sql = $db->prepare('INSERT INTO [Error] ([Timestamp], [IpAddress], [Url], [StatusCode]) VALUES (?, ?, ?, ?)');

    $sql->execute(array($timestamp, $address, $url, $code));

    $db = null;
  }
  catch(Exception $e)
  {
    // TODO: What now
  }
}

function get_footer_html()
{
  return '<footer>
  <p>Copyright &copy; 2018-2023 Cyotek Ltd. All Rights Reserved.</p>
  <p>System Version ' . get_system_version() . ' / Schema Version ' . get_schema_version() . '</p>
</footer></body></html>';
}

function get_header_html($title)
{
  return '<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0"> 
    <title>' . $title . '</title>
    <link rel="stylesheet" type="text/css" href="assets/styles.css" />
  </head>
  <body>';
}

function write_header_html($title)
{
  echo(get_header_html($title));
}

function write_footer_html()
{
  echo(get_footer_html());
}

function is_null_or_empty($value)
{
  return !(isset($value) && strlen($value) > 0);
}

function add_redirect($title, $slug, $url)
{
  $db = get_database_connection();

  $sql = 'INSERT INTO [Url] ([Enabled], [Title], [Slug], [Url]) VALUES (1, :title, :slug, :url);';

  $stmt = $db->prepare($sql);
  $stmt->bindValue(':title', $title);
  $stmt->bindValue(':slug', $slug);
  $stmt->bindValue(':url', $url);

  $stmt->execute();

  $db = null; 
}

function update_redirect($id, $enabled, $title, $slug, $url)
{
  $db = get_database_connection();

  $sql =  'UPDATE [Url]'
        . '   SET [Enabled] = :enabled'
        . '     , [Title]   = :title'
        . '     , [Slug]    = :slug'
        . '     , [Url]     = :url'
        . ' WHERE [Id]      = :id'
        ;

  $stmt = $db->prepare($sql);
  $stmt->bindValue(':id', $id);
  $stmt->bindValue(':enabled', $enabled);
  $stmt->bindValue(':title', $title);
  $stmt->bindValue(':slug', $slug);
  $stmt->bindValue(':url', $url);

  $stmt->execute();

  $db = null; 
}
?>
